File: traffic_monitor/state_manager.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import json
import logging
import os
from pathlib import Path
from typing import List, Dict, Any, Optional
from datetime import datetime


class StateManager:
    """
    Manages persistent state for the traffic monitor to avoid duplicate notifications
    after service restarts.
    """
    
    def __init__(self, state_file: Optional[str] = None):
        """
        Initialize the state manager.
        
        Args:
            state_file: Path to the state file. If None, uses default location.
        """
        self.logger = logging.getLogger(__name__)
        
        # Default state file location
        if state_file is None:
            state_dir = Path.home() / ".config" / "traffic-monitor"
            state_dir.mkdir(parents=True, exist_ok=True)
            self.state_file = state_dir / "state.json"
        else:
            self.state_file = Path(state_file)
            
        self.state_file.parent.mkdir(parents=True, exist_ok=True)
        
        # Initialize state
        self.state = self._load_state()
        
    def _load_state(self) -> Dict[str, Any]:
        """Load state from file."""
        if not self.state_file.exists():
            self.logger.debug(f"State file not found, creating new state: {self.state_file}")
            return self._create_default_state()
            
        try:
            with open(self.state_file, 'r') as f:
                state = json.load(f)
            self.logger.debug(f"Loaded state from {self.state_file}")
            return self._validate_state(state)
            
        except Exception as e:
            self.logger.error(f"Error loading state file {self.state_file}: {e}")
            self.logger.info("Creating new state file")
            return self._create_default_state()
    
    def _create_default_state(self) -> Dict[str, Any]:
        """Create default state structure."""
        return {
            "version": "1.0",
            "last_updated": datetime.now().isoformat(),
            "current_month": datetime.now().strftime("%Y-%m"),
            "notified_thresholds": [],
            "critical_notification_sent": False,
            "last_daily_report_date": None
        }
    
    def _validate_state(self, state: Dict[str, Any]) -> Dict[str, Any]:
        """Validate and migrate state if necessary."""
        current_month = datetime.now().strftime("%Y-%m")
        
        # If it's a new month, reset notification states
        if state.get("current_month") != current_month:
            self.logger.info(f"New month detected ({current_month}), resetting notification states")
            state["current_month"] = current_month
            state["notified_thresholds"] = []
            state["critical_notification_sent"] = False
            state["last_daily_report_date"] = None
            
        # Ensure all required fields exist
        default_state = self._create_default_state()
        for key, default_value in default_state.items():
            if key not in state:
                state[key] = default_value
                
        return state
    
    def _save_state(self) -> None:
        """Save current state to file."""
        try:
            self.state["last_updated"] = datetime.now().isoformat()
            
            with open(self.state_file, 'w') as f:
                json.dump(self.state, f, indent=2)
                
            self.logger.debug(f"Saved state to {self.state_file}")
            
        except Exception as e:
            self.logger.error(f"Error saving state to {self.state_file}: {e}")
    
    def get_notified_thresholds(self) -> List[int]:
        """Get list of thresholds that have already been notified."""
        return self.state.get("notified_thresholds", [])
    
    def add_notified_threshold(self, threshold: int) -> None:
        """Add a threshold to the notified list."""
        thresholds = self.state.get("notified_thresholds", [])
        if threshold not in thresholds:
            thresholds.append(threshold)
            thresholds.sort()  # Keep sorted for easier debugging
            self.state["notified_thresholds"] = thresholds
            self._save_state()
            self.logger.debug(f"Added threshold {threshold}GB to notified list")
    
    def is_critical_notification_sent(self) -> bool:
        """Check if critical notification has been sent."""
        return self.state.get("critical_notification_sent", False)
    
    def set_critical_notification_sent(self, sent: bool = True) -> None:
        """Set critical notification sent status."""
        self.state["critical_notification_sent"] = sent
        self._save_state()
        self.logger.debug(f"Set critical notification sent: {sent}")
    
    def get_last_daily_report_date(self) -> Optional[str]:
        """Get the date of the last daily report sent."""
        return self.state.get("last_daily_report_date")
    
    def set_last_daily_report_date(self, date_str: str) -> None:
        """Set the date of the last daily report sent."""
        self.state["last_daily_report_date"] = date_str
        self._save_state()
        self.logger.debug(f"Set last daily report date: {date_str}")
    
    def reset_monthly_state(self) -> None:
        """Reset state for a new month (useful for testing or manual reset)."""
        current_month = datetime.now().strftime("%Y-%m")
        self.state["current_month"] = current_month
        self.state["notified_thresholds"] = []
        self.state["critical_notification_sent"] = False
        self.state["last_daily_report_date"] = None
        self._save_state()
        self.logger.info("Reset monthly notification state")
    
    def get_state_summary(self) -> Dict[str, Any]:
        """Get a summary of current state for debugging."""
        return {
            "state_file": str(self.state_file),
            "current_month": self.state.get("current_month"),
            "notified_thresholds": self.state.get("notified_thresholds", []),
            "critical_notification_sent": self.state.get("critical_notification_sent", False),
            "last_daily_report_date": self.state.get("last_daily_report_date"),
            "last_updated": self.state.get("last_updated")
        }

File: src/state_manager.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};
use chrono::Local;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct State {
    pub version: String,
    pub last_updated: String,
    pub current_month: String,
    pub notified_thresholds: Vec<u64>,
    pub critical_notification_sent: bool,
    pub last_daily_report_date: Option<String>,
}

impl Default for State {
    fn default() -> Self {
        let now = Local::now();
        Self {
            version: "1.0".to_string(),
            last_updated: now.to_rfc3339(),
            current_month: now.format("%Y-%m").to_string(),
            notified_thresholds: Vec::new(),
            critical_notification_sent: false,
            last_daily_report_date: None,
        }
    }
}

pub struct StateManager {
    state_file: PathBuf,
    state: State,
}

fn get_home_dir() -> Option<PathBuf> {
    std::env::var("HOME")
        .or_else(|_| std::env::var("USERPROFILE"))
        .ok()
        .map(PathBuf::from)
}

impl StateManager {
    pub fn new(state_file: Option<PathBuf>) -> Result<Self, Box<dyn std::error::Error>> {
        let path = match state_file {
            Some(p) => p,
            None => {
                let home = get_home_dir().ok_or("Cannot find home directory")?;
                home.join(".config").join("traffic-monitor").join("state.json")
            }
        };

        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        let mut manager = Self {
            state_file: path,
            state: State::default(),
        };

        manager.load_state()?;
        Ok(manager)
    }

    fn load_state(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        if !self.state_file.exists() {
            self.state = State::default();
            self.save_state()?;
            return Ok(());
        }

        let content = fs::read_to_string(&self.state_file)?;
        match serde_json::from_str::<State>(&content) {
            Ok(loaded_state) => {
                self.state = self.validate_state(loaded_state);
            }
            Err(_) => {
                self.state = State::default();
                self.save_state()?;
            }
        }
        Ok(())
    }

    fn validate_state(&self, mut state: State) -> State {
        let now = Local::now();
        let current_month = now.format("%Y-%m").to_string();

        if state.current_month != current_month {
            state.current_month = current_month;
            state.notified_thresholds.clear();
            state.critical_notification_sent = false;
            state.last_daily_report_date = None;
        }

        state
    }

    fn save_state(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        self.state.last_updated = Local::now().to_rfc3339();
        let content = serde_json::to_string_pretty(&self.state)?;
        fs::write(&self.state_file, content)?;
        Ok(())
    }

    pub fn get_notified_thresholds(&self) -> &[u64] {
        &self.state.notified_thresholds
    }

    pub fn add_notified_threshold(&mut self, threshold: u64) -> Result<(), Box<dyn std::error::Error>> {
        if !self.state.notified_thresholds.contains(&threshold) {
            self.state.notified_thresholds.push(threshold);
            self.state.notified_thresholds.sort();
            self.save_state()?;
        }
        Ok(())
    }

    pub fn is_critical_notification_sent(&self) -> bool {
        self.state.critical_notification_sent
    }

    pub fn set_critical_notification_sent(&mut self, sent: bool) -> Result<(), Box<dyn std::error::Error>> {
        self.state.critical_notification_sent = sent;
        self.save_state()?;
        Ok(())
    }

    pub fn get_last_daily_report_date(&self) -> Option<&str> {
        self.state.last_daily_report_date.as_deref()
    }

    pub fn set_last_daily_report_date(&mut self, date_str: String) -> Result<(), Box<dyn std::error::Error>> {
        self.state.last_daily_report_date = Some(date_str);
        self.save_state()?;
        Ok(())
    }

    pub fn reset_monthly_state(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        let now = Local::now();
        self.state.current_month = now.format("%Y-%m").to_string();
        self.state.notified_thresholds.clear();
        self.state.critical_notification_sent = false;
        self.state.last_daily_report_date = None;
        self.save_state()?;
        Ok(())
    }

    pub fn get_state(&self) -> &State {
        &self.state
    }

    pub fn get_state_file_path(&self) -> &Path {
        &self.state_file
    }
}
